clang/test/SemaSYCL/sycl-force-inline-kernel-lambda-ast.cpp

```diff
@@ -0,0 +1,38 @@
+// RUN: %clang_cc1 -fno-sycl-force-inline-kernel-lambda -fsycl-is-device -internal-isystem %S/Inputs -disable-llvm-passes -triple spir64-unkown-unknown -ast-dump -o - %s | FileCheck %s
+// RUN: %clang_cc1 -fsycl-is-device -internal-isystem %S/Inputs -disable-llvm-passes -triple spir64-unkown-unknown -ast-dump -o - %s | FileCheck %s
+
+#include "sycl.hpp"
```

Comment:
Why did you change the IR test to an AST test?

Reply:
Because tests in `SemaSYCL` should not be checking LLVM IR. And the fact that the only thing `-f[no-]sycl-force-inline-kernel-lambda` flag does is add (or not) an attribute at AST level. 

Unit-test should check the exact change we want to test in maximum isolation from other components. Strictly speaking, there is no way to check `-f[no-]sycl-force-inline-kernel-lambda` flag at LLVM IR level, because inlining of a kernel lambda may happen in optimization pass anyway regarless of that flag.

Reply:
> Both RUN lines are checking the same thing.

Good point.

> Can you elaborate on what you mean by -f[no-]sycl-force-inline-kernel-lambda flag adds an attribute or not?

From what I see in b91b732783fb1cabac6652945deb9e6662480960, `AlwaysInlineAttr` should be added conditionally and it can be controlled with a flag. Perhaps we are checking wrong AST entity, because I would expect to see `AlwaysInlineAttr` attached somewhere for the first kernel (`KernelName`)

Reply:
This was the only conceivable way to perform the test as in the AST, the kernel definition was separated from the lambda expression, and the inlining attributes were placed on/in the `CXXMethodDecl` node in the `LambdaExpr` nodes themselves, rather than on the kernel definition. The `FunctionDecl` and `LambdaExpr` nodes could be linked by their hexadecimal address but I found no facility in the testing functions - CHECK etc. - to match these up.

I would have liked to have linked them up with the actual kernel definitions and names but that didn't seem to be an option here.

Reply:
You can do - 

```
q.submit([&](sycl::handler &h) {
    // CHECK: LambdaExpr{{.*}}sycl-force-inline-kernel-lambda-ast.cpp:13
    // INLINE: AlwaysInlineAttr
    // NOINLINE-NOT: AlwaysInlineAttr
    h.parallel_for<class KernelName>([] {});
  });

  q.submit([&](sycl::handler &h) {
    // CHECK: LambdaExpr{{.*}}sycl-force-inline-kernel-lambda-ast.cpp:19
    // CHECK: AlwaysInlineAttr
    h.parallel_for<class KernelNameInline>([]() __attribute__((always_inline)) {});
  });

  q.submit([&](sycl::handler &h) {
    // CHECK: LambdaExpr{{.*}}sycl-force-inline-kernel-lambda-ast.cpp:26
    // CHECK: NoInlineAttr
    // CHECK-NOT: AlwaysInlineAttr
    h.parallel_for<class KernelNameNoInline>([]() __attribute__((noinline)) {});
  });

  /// The flag is ignored for ESIMD kernels
  q.submit([&](sycl::handler &h) {
    // CHECK: LambdaExpr{{.*}}sycl-force-inline-kernel-lambda-ast.cpp:35
    // CHECK: SYCLSimdAttr
    // CHECK-NOT: AlwaysInlineAttr
    // CHECK-NOT: NoInlineAttr
    h.parallel_for<class KernelNameESIMD>([]() __attribute__((sycl_explicit_simd)) {});
  });
```
The corresponding RUN lines would be - 

```
// RUN: %clang_cc1 -fno-sycl-force-inline-kernel-lambda -fsycl-is-device -internal-isystem %S/Inputs -disable-llvm-passes -triple spir64-unknown-unknown -ast-dump -o - %s | FileCheck %s --check-prefixes=NOINLINE,CHECK
// RUN: %clang_cc1 -fsycl-is-device -internal-isystem %S/Inputs -disable-llvm-passes -triple spir64-unknown-unknown -ast-dump -o - %s | FileCheck %s --check-prefixes=INLINE,CHECK
```

Reply:
Thank you for your help with this. I tried using FileCheck's string substitution blocks but found picking out the correct information from the appropriate AST nodes and combining them was a lot trickier and more verbose than this approach.

Reply:
No problem!